doc/arkode/guide/source/Usage/MRIStep/MRIStepCoupling.rst

```diff
@@ -57,22 +82,32 @@ constructing a coupling table and attaching it with
 
       The embedding order of accuracy
 
+   .. c:member:: sunrealtype* c
+
+      An array of length ``[stages]`` containing the slow abscissae :math:`c^S`
+      for the method
+
    .. c:member:: sunrealtype*** W
 
-      A three-dimensional array with dimensions ``[nmat][stages][stages]``
+      A three-dimensional array with dimensions ``[nmat][stages+1][stages]``
       containing the method's :math:`\Omega^{\{k\}}` coupling matrices for the
       slow-nonstiff (explicit) terms in :eq:`ARKODE_IVP_two_rate`
 
    .. c:member:: sunrealtype*** G
 
-      A three-dimensional array with dimensions ``[nmat][stages][stages]``
+      A three-dimensional array with dimensions ``[nmat][stages+1][stages]``
       containing the method's :math:`\Gamma^{\{k\}}` coupling matrices for the
       slow-stiff (implicit) terms in :eq:`ARKODE_IVP_two_rate`
 
-   .. c:member:: sunrealtype* c
+   .. c:member:: int ngroup
 
-      An array of length ``[stages]`` containing the slow abscissae :math:`c^S`
-      for the method
+      Number of stage groups for the method (only relevant for MERK methods).
+
+   .. c:member:: int** group
+
+      A two-dimensional array with dimensions ``[stages][stages]`` that encodes
```

Comment:
The `stages x stages` storage is an upper bound on the actually-required storage for these stage groups. 
 The maximum number of groups is `stages`, and the maximum number of entries in any group is `stages`. 
 However, since each group can have differing numbers of entries, then the cleanest way for the `MRIStepCoupling_Alloc` to allocate space for these is to use this upper bound.

Internally, I use a value of `-1` to indicate the end of a group, which is useful because each group can have a different size.

Regarding your hope of deducing the stage groups automatically using sparsity -- deduction would require both sparsity *and* floating-point comparison operations to ensure that the forcing coefficients for "grouped" stages were identical (note: the whole point of a "group" is that their fast IVPs use identical initial conditions and forcing functions but have different interval stopping points, so by grouping we don't repeat the shared fast interval integrations).  If I trusted floating-point comparisons then perhaps deduction would make sense, but I don't trust them.

While use of the `stages x stages` structure wastes some storage, that waste is insignificant -- the worst-case method is MERK54, that has 11 stages with 5 "active" groups.  Here, the current format requires 121 integers, whereas a sparse format would require 27 integers (plus a lot more headache and potential bugs).  Thus for this largest MERK method we only waste storage of 94 integers (376 bytes).

Reply:
It would be good to document this padding, what `-1` does, and the encoding format.

I understand the concern for floating point comparison, but I feel more optimistic that it can be done safely. For me, the main issue is where does the computation to deduce the groups belong. We don't want it in the take step function.

I'll throw out another idea for the group format (if I understand things correctly). A single array with `ngroup` elements (or `stages` to pad to the maximum). Each entry represents how many stages are in the group. For example [2, 3, 1] means stages 1,2 are in group 1, stages 3,4,5 are in group 2, and stage 6 is in group 3. This requires stages within the same group to be consecutive but that feels like a small restriction.

Reply:
Unfortunately, the "consecutive" assumption is not how these were originally invented -- for example, with MERK54 the first group evolves stage 1, then 3, then 2; the second group evolves stage 6, then 5, then 4, etc.  While it is certainly possible to reorder everything in the coefficient tables so that the stage groups are consecutive, there would be a huge "disconnect" between our coefficients and the published values, and I worry that it would seriously complicate future understanding by users/collaborators/new students/new postdocs.

If we did add a "deduction" phase then it could safely be added into `MRIStepSetCoupling` to deduce values input from a user, and into `mriStep_SetCoupling` just after the default table for a given order and method type is loaded.  That said, I still prefer the more straightforward [current] approach, along with extra documentation to explain how a user could set up their own groups for custom MERK methods.

Reply:
I think the "deduction" is something that can be a separate enhancement that we add to our backlog as it seems like too big of a feature to add-on this late.